Mass transit (part 2 of 2)

yjmrobert opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->

## Description
<!-- Please provide a brief description of the changes made in this pull request -->
Leveraged exploratory work from @vb-ssc  and @ErikApption in https://github.com/ssc-sp/datahub-portal/pull/1076 to convert the ServerlessOperations azure functions to use azure service bus. Also had some minor bug fixes that were caught by updating unit tests.

> _Note: This is part one of two, the portal and resource provisioner still need to communicate off the service bus_

## How Has This Been Tested?
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->
Ran pings through the functions triggering off the infra dev service bus queues. Additionally passing all unit tests locally.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Unit test (new or update to tests)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (change that would cause documentation to be updated)
- [x] Configuration change (change that would affect the configuration of the application)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask! -->
- [x] PR is submitted to the correct branch `develop`.
- [x] Code follows the code style of this project.
- [x] Changes are covered by Specflow tests and all new or existing tests are passing.
- [x] Any new or updated translatable strings have been added to the localization files.
- [x] Necessary and relevant documentation has been created or updated.
- [x] There are either no changes to the application's configuration or the necessary changes in the infrastructure repository are included in a separate PR.

<!-- Link to the "infrastructure" repository PR here -->

Mass transit requires the service bus queues to exist when spinning up the functions. Manually created the queues and updated the infrastructure terraform to match desired end state.

https://github.com/ssc-sp/datahub-infra/pull/88

yjmrobert committed:
🔧 Add DatahubServiceBus and update QueueNames configurations

A new configuration setting, DatahubServiceBus, has been added to DatahubPortalConfiguration. Additionally, QueueNames in the QueueNames class has been updated with new entries for "DatabricksSyncOutput", "KeyvaultSyncOutput" and "StorageSyncOutput".

yjmrobert committed:
🔧 Added Mass Transit to serverless operations

yjmrobert committed:
🔧 Moved BugReport over to MT and refactored service bus queue names

yjmrobert committed:
🔧 Added queue constants

yjmrobert committed:
🔧 Added extension method for mediator pattern on service bus

yjmrobert committed:
🔧 Updated infrastructure status queue to service bus

yjmrobert committed:
🔧 Updated workspace app service queue to service bus

yjmrobert committed:
🔧 Removed mediatr from serverless operations

yjmrobert committed:
🔧 Updated email notification handler to service bus

yjmrobert committed:
🔧 Updated project usage notifier to service bus

yjmrobert committed:
🔧 Updated inactivity queue and refactored pong service into infrastructure along with messaging extension

yjmrobert committed:
🔧 Updated project usage queue and added cancellation token to extension method

yjmrobert committed:
🔧 update project usage function to service bus

yjmrobert committed:
🔧 update tf output handler to service bus

yjmrobert committed:
🐛 fixed name

yjmrobert committed:
🔧 updated user inactivity to service bus

yjmrobert committed:
🔧 added connection string for service bus

yjmrobert committed:
🔧 Updated connection string key in ConfigureServices

The key used to retrieve the DatahubServiceBus connection string in the ConfigureServices file was incorrectly read as "DatahubServiceBus". This commit modifies the key to the correct "DatahubServiceBus__ConnectionString". Now, an ArgumentNullException will be properly thrown if this key-value pair does not exist in configuration.

yjmrobert committed:
🔧 Update .gitignore in Datahub.Functions

Added appsettings.json to the .gitignore file in the Datahub.Functions directory. This update will prevent the accidental tracking and committing of sensitive information found in the appsettings.json file.

yjmrobert committed:
🔧 Refactor serverless operations for ServiceBus integration

This commit pertains to refactoring operations on the Serverless side of the application to integrate with the Azure ServiceBus. The changes include modifying the parameters of several methods to include a connection string connected to the ServiceBus. Furthermore, the Mediator library was replaced with MassTransit for better support of Azure ServiceBus. There were also various small formatting and stylistic changes made for improved readability.

yjmrobert committed:
🐛 fixed all failing unit tests

yjmrobert committed:
🔧 Update Service Bus Connection String in Config Files

The Service Bus connection string setting was updated in three configuration files: template.settings.json in ResourceProvisioner, template.appsettings.json in Portal and template.settings.json in ServerlessOperations. 'MassTransit' settings were replaced with 'DatahubServiceBus' which directly uses a connection string from the KeyVault.

yjmrobert committed:
🔧 Integrate MassTransit and update PackageReferences

Added MassTransit to the Program.cs and updated Azure.Identity version in the csproj file. Also introduced the ServiceBusTrigger replacing the previous QueueTrigger, requiring a new queue name constant and changes in the deserialization of the resource run request. These updates enhance the service bus processing capabilities.

yjmrobert committed:
Merge branch 'refs/heads/develop' into yr/mass-transit

yjmrobert committed:
🔧 Update configuration settings and cleanup csproj

Updated the application configuration settings in template.settings.json and Program.cs for Azure Functions, and set local.settings.json to be non-optional. Cleaned up ResourceProvisioner.Functions.csproj by removing redundant configuration files.

yjmrobert committed:
🔧 Refactor queue message sender and simplify project usage update

Removed unnecessary configurations in the queue message sender, introduced MassTransit for sending service bus messages and simplified the project usage update code by removing a timeout parameter that was not in use. These changes aim at improving code simplicity, readability, and efficiency.

yjmrobert committed:
🔧 Renamed health check queue sender

yjmrobert committed:
🔧 Implement IDisposable in WelcomeBanner

Added IDisposable interface to the WelcomeBanner class along with a Dispose method. This in addition to the existing IAsyncDisposable interface, to increase disposal control and flexibility. This provides an additional option to free up unmanaged resources synchronously.

yjmrobert committed:
🔧 Refactor messaging service to use MassTransit

This update includes changes to the ResourceMessagingService, which no longer directly interacts with the AzureStorageQueue for sending messages. Instead, it now depends on the MassTransit ISendEndpointProvider interface. The specifics about the queue (like names and connection strings) have been moved to a shared location, improving modularity and flexibility across the codebase.

yjmrobert committed:
🐛 bugfix

yjmrobert committed:
🔧 Refactored ResourceRunRequest to use IRepositoryService

In this update, the implementation of the class ResourceRunRequest has been changed to use IRepositoryService instead of CreateResourceRunCommandHandler. This alteration provides a more robust handling of the ResourceRun object and enables logging of pull request messages successfully processed.

yjmrobert committed:
🔧 Add ServiceBusReceivedMessage extension to deserialize messages

Refactored codebase to use ServiceBusReceivedMessage extension for deserializing ServiceBus messages. This extension handles case-insensitive property deserialization and message unwrapping, simplifying the main functional components. As a result, redundant deserialization code was removed from multiple serviced functions across the solution which not only improved the code readability but also standardized the message processing method.

yjmrobert committed:
🔧 fixed unit tests

ErikApption commented on Portal/src/Datahub.Infrastructure/Datahub.Infrastructure.csproj:
```diff
@@ -14,6 +14,9 @@
       <PackageReference Include="Azure.ResourceManager.Storage" Version="1.2.0" />
       <PackageReference Include="Google.Cloud.Storage.V1" Version="4.6.0" />
+      <PackageReference Include="MassTransit" Version="8.2.2" />
+      <PackageReference Include="MassTransit.Abstractions" Version="8.2.2" />
```
we can prob ignore the abstraction and mass transit packages

yjmrobert replied:
weird, not sure how that made it in

ErikApption commented on Portal/src/Datahub.Portal/Startup.cs:
```diff
@@ -102,7 +102,10 @@ public Startup(IConfiguration configuration, IWebHostEnvironment env)
     public void ConfigureServices(IServiceCollection services)
     {
-        services.AddApplicationInsightsTelemetry();
+        services.AddApplicationInsightsTelemetry(x =>
```
we prob should ensure we are consistent with app insights in the other modules and use the conn string instead of key

yjmrobert replied:
are you suggesting we add application insights to the serverless operations and resource provisioner?

this is the only instance of it being used

ErikApption replied:
rather thinking that we kill the key from TF and only rely on conn string

yjmrobert replied:
ah I see, yes the TF needs to be updated to set the connection string to `ApplicationInsights:ConnectionString`

yjmrobert committed:
🔧 removed abstractions package

yjmrobert committed:
🔧 Add alphanumeric resource prefix

This commit adds an alphanumeric resource prefix in the 'template.settings.json' and 'appsettings.psm1' files. The alphanumeric resource prefix was created to ensure compatibility with Azure resources which may not support certain special characters.
